cvmfs/swissknife_lease.cc

```diff
@@ -0,0 +1,141 @@
+/**
+ * This file is part of the CernVM File System.
+ */
+
+#include "swissknife_lease.h"
+
+#include <fstream>
+#include <sstream>
+
+#include "swissknife_lease_curl.h"
+#include "swissknife_lease_json.h"
+
+#include "logging.h"
+
+namespace {
+
+bool CheckParams(const swissknife::CommandLease::Parameters& p) {
+  if (p.action != "acquire" && p.action != "drop") {
+    return false;
+  }
+
+  return true;
+}
+
+}  // namespace
+
+namespace swissknife {
+
+enum LeaseError {
+  LEASE_SUCCESS           = 0,
+  LEASE_UNUSED            = 1,
+  LEASE_PARAM_ERROR       = 2,
+  LEASE_CURL_INIT_ERROR   = 3,
+  LEASE_FILE_OPEN_ERROR   = 4,
+  LEASE_FILE_DELETE_ERROR = 5,
+  LEASE_PARSE_ERROR       = 6,
+  LEASE_CURL_REQ_ERROR    = 7,
+};
+
+CommandLease::~CommandLease() {
+}
+
+ParameterList CommandLease::GetParams() const {
+  ParameterList r;
+  r.push_back(Parameter::Mandatory('u', "repo service url"));
+  r.push_back(Parameter::Mandatory('a', "action (acquire or drop)"));
+  r.push_back(Parameter::Mandatory('n', "user name"));
+  r.push_back(Parameter::Mandatory('p', "lease path"));
+  return r;
+}
+
+int CommandLease::Main(const ArgumentList& args) {
+  Parameters params;
+
+  params.repo_service_url = *(args.find('u')->second);
+  params.action = *(args.find('a')->second);
+  params.user_name = *(args.find('n')->second);
+  params.lease_fqdn = *(args.find('p')->second);
+
+  if (!CheckParams(params)) {
+    return LEASE_PARAM_ERROR;
+  }
+
+  // Initialize curl
+  if (curl_global_init(CURL_GLOBAL_ALL)) {
+    return LEASE_CURL_INIT_ERROR;
+  }
+
+  LeaseError ret = LEASE_SUCCESS;
+  if (params.action == "acquire") {
+    CurlBuffer buffer;
+    if (MakeAcquireRequest(params.user_name,
+                           params.lease_fqdn,
+                           params.repo_service_url,
+                           &buffer)) {
+      std::string session_token;
+      if (ParseAcquireReply(buffer, &session_token)) {
+        // Save session token to /var/spool/cvmfs/<REPO_NAME>
+        // TODO(radu): Is there a special way to access the scratch directory?
+        std::string token_file_name = "/var/spool/cvmfs/" +
+            params.lease_fqdn + "/session_token";
```

Comment:
I think we should be prepared to have multiple concurrent sessions on the same release manager machine, e.g. by adding a seesion uuid to the file name